# DLL Lab — 22 January 2026
## Single Layer Perceptron with Gradient Descent
**Done By:** Student | **Course:** AI-612


In [1]:
import tensorflow.compat.v1 as tf
tf.disable_v2_behavior()
import numpy as np
import matplotlib.pyplot as plt
import warnings; warnings.filterwarnings('ignore')
print('Libraries loaded.')


Libraries loaded.


## Define Model

In [2]:
w = tf.Variable([.2], dtype=tf.float32)
b = tf.Variable([-.2], dtype=tf.float32)
x = tf.placeholder(tf.float32, name='x')
y = tf.placeholder(tf.float32, name='y')
linear_model = w * x + b
print('Model: y_hat = w*x + b')
print('Initial w:', [0.2], '  Initial b:', [-0.2])


Model: y_hat = w*x + b
Initial w: [0.2]   Initial b: [-0.2]


## Loss and Optimizer

In [3]:
squared_deltas = tf.square(linear_model - y)
loss = tf.reduce_sum(squared_deltas)
optimizer = tf.train.GradientDescentOptimizer(0.01)
train = optimizer.minimize(loss)
print('Loss: Sum of Squared Errors')
print('Optimizer: Gradient Descent  lr=0.01')


Loss: Sum of Squared Errors
Optimizer: Gradient Descent  lr=0.01


## Train for 1000 Epochs

In [4]:
sess = tf.Session()
sess.run(tf.global_variables_initializer())

x_data = [1, 2, 3, 4]
y_data = [0, -1, -2, -3]

loss_before = sess.run(loss, {x: x_data, y: y_data})
print(f'Loss before training: {loss_before:.4f}')

loss_history = []
for i in range(1000):
    sess.run(train, {x: x_data, y: y_data})
    loss_history.append(sess.run(loss, {x: x_data, y: y_data}))

loss_after = sess.run(loss, {x: x_data, y: y_data})
final_w, final_b = sess.run([w, b])
print(f'Loss after  training: {loss_after:.2e}')
print(f'Final w: {final_w}  Final b: {final_b}')


Loss before training: 20.1600
Loss after  training: 4.66e-11
Final w: [-0.99999714]  Final b: [0.99999166]


## Predictions

In [5]:
preds = sess.run(linear_model, {x: x_data, y: y_data})
print(f'{'x':>4}  {'y_true':>8}  {'y_pred':>10}')
print('-'*26)
for xi, yi, yp in zip(x_data, y_data, preds):
    print(f'{xi:>4}  {yi:>8}  {yp:>10.5f}')


   x    y_true      y_pred
--------------------------
   1         0    -0.00001
   2        -1    -1.00000
   3        -2    -2.00000
   4        -3    -2.99997


## Loss Curve

In [6]:
plt.figure(figsize=(7, 4))
plt.plot(loss_history, color='steelblue', linewidth=2)
plt.yscale('log')
plt.title('Training Loss over 1000 Epochs', fontsize=13)
plt.xlabel('Epoch'); plt.ylabel('Loss (log scale)')
plt.grid(True, alpha=0.3)
plt.tight_layout()
plt.show()
print('Loss curve plotted.')


Loss curve plotted.


## Summary
| | Value |
|---|---|
| Loss before | 20.16 |
| Loss after 1000 epochs | 4.66e-11 |
| Final w | ≈ -1.0 |
| Final b | ≈ +1.0 |

Gradient Descent successfully converged the perceptron to the true relationship **y = -x + 1**.